src/utils/fmt.zig

```diff
@@ -1,38 +1,240 @@
-            const ptr = &@field(max_args, field.name);
-            ptr.* = maxArg(field);
-        }
-        return max_args;
-    }
-}
-
+/// Returns the maximum length applicable for the format string and `Args` tuple,
+/// such that it would be the equivalent to the length of the bounded array returned
+/// by `boundedFmt`.
+/// ```zig
+/// try expectEqual("255-1".len, boundedLen("{d}-{d}", struct { u8, u1 }));
+/// // comptime field values in the struct/tuple are reflected appropriately
+/// try expectEqual("foo-255".len, boundedLen("{[a]s}-{[b]d}", struct { comptime a: []const u8 = "foo", b: u8 }));
+/// ```
 pub inline fn boundedLen(
     comptime fmt_str: []const u8,
     comptime Args: type,
 ) usize {
     comptime return std.fmt.count(fmt_str, maxArgs(Args));
 }
 
+/// Same as `boundedLen`, but takes a value instead of a type.
+/// The values should posses the maximum values applicable for each
+/// element type, in order to match the actual bounded length.
+/// For example:
+/// ```zig
+/// try expectEqual("255-255".len, boundedLenValue("{d}-{d}", .{ std.math.maxInt(u8), std.math.maxInt(u8) }));
+/// // comptime field values in the struct/tuple are reflected appropriately
+/// try expectEqual("foo-255".len, boundedLenValue("{[a]s}-{[b]d}", .{ .a = "foo", .b = 255 }));
+/// ```
+pub inline fn boundedLenValue(
+    comptime fmt_str: []const u8,
+    comptime value: anytype,
+) usize {
+    comptime return boundedLen(fmt_str, @TypeOf(value));
+}
+
 /// Returns a bounded array string, guaranteed to be able to represent the formatted result.
-pub fn boundedFmt(
+/// For example:
+/// ```zig
+/// try expectEqualStrings("fizz.buzz", boundedFmt("{s}.{s}", .{ "foo", "buzz" }).constSlice());
+/// ```
+pub inline fn boundedFmt(
     comptime fmt_str: []const u8,
     args: anytype,
 ) std.BoundedArray(u8, boundedLen(fmt_str, @TypeOf(args))) {
     var result: std.BoundedArray(u8, boundedLen(fmt_str, @TypeOf(args))) = .{};
     result.writer().print(fmt_str, args) catch unreachable;
     return result;
 }
+
+/// Returns an instance of the tuple of type `Args`, wherein each
+/// element of the tuple possesses the maximum value applicable
+/// to the type of the element.
+/// For example:
+/// ```zig
+/// try expectEqual();
+/// ```
+pub inline fn maxArgs(comptime Args: type) Args {
+    comptime {
+        var max_args: Args = undefined;
+        for (@typeInfo(Args).Struct.fields) |field| {
+            if (field.is_comptime) continue;
+            const ptr = &@field(max_args, field.name);
+            ptr.* = maxArg(field.type);
+        }
+        return max_args;
+    }
+}
+
+/// Returns a wrapper around the bounded array which will be usable as an argument
+/// to `boundedFmt` and `boundedLenValue`.
+pub inline fn boundedString(
+    /// `*const std.BoundedArray(u8, capacity)`
+    bounded: anytype,
+) if (sig.utils.types.boundedArrayInfo(@TypeOf(bounded.*))) |ba_info| BoundedString(ba_info.capacity) else noreturn {
```

Comment:
I don't think this needs an `if` either (like my other comments). But here I guess it might circumvent your custom compile error.

Reply:
Indeed, compile error in the signature isn't ideal; as described in my other comment, that can lead to poor compiler error stack traces.